[Feature] roctracer integration.

whchung opened the pull request:
Rewrite PR #455 to cope with latest development in TF mainline and roctracer.

Captures kernel invocations / memory transfers / memory allocations.

- mGPU trace collection is implemented per following rule:
  a) within an API callback, associated device ID and stream ID is not known.
  b) within an activity callback, associated device ID and stream ID is not known at HIP domain.
  c) within an activity callback, associated device ID and stream ID is known at HCC domain.
  Based on the information collected in a) , b), c). It's possible to associate all three and form the proper map. `StepStatsRocmTracerAdaptor` has been revised to gradually construct all relevant information.


Some caveats yet resolved.

- `roctracer-dev` is assumed to be present in default ROCm installation.
- `-lroctracer64` is explicitly added to `crosstool`.
- Quite a few duplicated logic between `rocm_tracer` and its CUPTI counterpart.
- Stream synchronization events are yet captured.
- P2P memory transfer destination is not logged.
- meaning of certain timing fields needs to be reviewed.

Refer to `ROCM TODO` in the changeset for more information on known caveats.

Tests carried out:
```
 bazel test --config=rocm //tensorflow/core/profiler/internal/gpu:device_tracer_test
```

Need application-level testing / tracing as well.

whchung committed:
Initial commit enabling roctracer on TensorFlow.

whchung committed:
Enable device_tracer_test on ROCm.

whchung committed:
Populate rocm_tracer.

whchung committed:
Amend device_tracer_rocm to pick up rocm_tracer.

whchung committed:
Make device_tracer_test run to end without crashing.

whchung committed:
Initial commit to hook up with roctracer.

whchung committed:
Getting API and activity callbacks populated into TF profiler data structures.

deven-amd commented on tensorflow/core/profiler/internal/gpu/rocm_tracer.h:
```diff
@@ -0,0 +1,256 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
```
playing devils advocate here

is there way we can have a common `gpu_tracer.h`? 
* Then for the implementation we can have `cupti_wrapper.cc` on the cuda side, and `rocm_tracer.cc` on the rocm side
* This is would also potentially allow us to keep a common `device_tracer.cc` file

Think it would be very difficult to pull this off, but odds are this question will come up during the upstream review process.

whchung replied:
@deven-amd we certainly can, and probably should extract common logic for upstreaming purpose...

I just revised HCC patch ( https://github.com/RadeonOpenCompute/hcc/pull/1314 ) after some discussion with @scchan and @jeffdaily . Now working on getting `rocm_tracer.cc` to incorporate mGPU book-keeping logic. Will then address comments from you here.

whchung commented on third_party/gpus/rocm_configure.bzl:
```diff
@@ -710,6 +721,12 @@ def _create_local_rocm_repository(repository_ctx):
             src_dir = rocm_toolkit_path + "/rccl/include",
             out_dir = "rocm/include/rccl",
         ),
+        #make_copy_dir_rule(
```
I don't quite recall, but it has to do with the fact we explicitly `-lroctracer64` for now.

whchung commented on tensorflow/core/profiler/internal/gpu/BUILD:
```diff
@@ -28,6 +30,7 @@ tf_cuda_library(
     cuda_deps = [
         "//tensorflow/core/profiler/internal/gpu:cupti_tracer",
         "//tensorflow/core/profiler/internal/gpu:cupti_wrapper",
+        "//tensorflow/core/profiler/internal/gpu:rocm_tracer",
```
doing so means adding quite a few boilerplate codes inside `StreamExecutor`. probably we should still need to do that in case we want to get rid of `-lroctracer64` within `crosstool`.

I'd like to revisit this after I get `rocm_tracer` completely mGPU-aware.

whchung committed:
Rename classes.

RocmDriverApiHook -> RocmApiHook.
RocmDriverApiHookWithActivityApi -> RocmApiHookImpl.

whchung committed:
Revise API callback data structures and interfaces.

device_id is really not available in HIP API callbacks.

whchung committed:
Remove device_id from AnnotationMap.

whchung commented:
Pushed a couple of more commits to lay the infrastructure for mGPU support, and to cope with behavioral differences between `CUPTI` and `roctracer`. In API callbacks, `CUPTI` has `device_id` and `stream_id`, but not in `roctracer` API callbacks.

Such information is only available at activity callbacks within HCC domain. The information could be reconstructed traversing `correlation_id` field.

Also when roctracer invokes activity callbacks for HCC domain has also to be changed. That has been addressed in the patch for HCC PR#1314.

In the next couple of commits such reconstruction would be introduced.

whchung committed:
Heavily revise data reporting logic.

whchung committed:
enabling profiler/tracer tests for the ROCm platform

whchung committed:
Compute gpu start time with roctracer_get_timestamp API.

whchung committed:
XXX: ROCM-specific changes to make chrome tracing looks pretty.

ROCm-Apps-Test commented:
Can one of the admins verify this patch?

sunway513 commented:
@whchung , can you rebase the PR with the tip of develop-upstream?
